### ISSUE TITLE
add configurable limit to the kernel runtime 

### ISSUE THREAD

**Issue by rochaporto** [jupyterhub/jupyterhub#2598]
We rely on jupyterhub for a few multi user notebook deployments, backed by Kubernetes and exposing different sets of resources.

In cases where the resources are limited (as is the case for GPUs), it would be great to be able to configure a max runtime for the user server so that everyone gets their share. This is similar to what kaggle offers with session limits.

We rely on the zero2jupyterhub helm chart to setup and config the server, so maybe a good option would be to add this in the profile definition.

**Comment by rochaporto** [comment]
Culling seems to be doing already what was described above:
https://github.com/jupyterhub/jupyterhub/blob/master/examples/cull-idle/cull_idle_servers.py#L331

Apart from the separation between different server types: cpu only, cpu+gpu, etc. Would be nice to set different maxAge values for each of these. Is this already possible somehow? I couldn't figure this out from the code and config, maybe it needs an additional param to set the server profile(s) the cull service should aim for? Could then deploy more than one, with different configs.

**Comment by vilhelmen** [comment]
I don't think there's anything in place to support this. I'll take a closer look, though. One route would be to attach custom `maxAge` and `maxIdle` to spawner objects that the culler can receive through the API, but that seems annoyingly complicated. A better method would be making the culler more configurable and allowing users to map some sort of container name to lifetimes. It all depends on what the culler can get through the API.

**Comment by vilhelmen** [comment]
Yeah, the culler currently only receives the following snippets of spawner data from `/users`, it would need to be extended to transmit some sort of container id data:

```
{
    "kind": "user",
    "name": "XXX",
    "admin": false,
    "groups": [],
    "server": "/user/XXX/",
    "pending": null,
    "created": "2019-06-05T19:02:28.458702Z",
    "last_activity": "2019-06-05T19:35:56.697797Z",
    "servers": {
      "": {
        "name": "",
        "last_activity": "2019-06-05T19:35:56.633000Z",
        "started": "2019-06-05T19:02:36.859734Z",
        "pending": null,
        "ready": true,
        "state": {
          "pod_name": "jupyter-XXX"
        },
        "url": "/user/XXX/",
        "progress_url": "/hub/api/users/XXX/server/progress"
      }
    }
```

**Comment by vilhelmen** [comment]
After whipping up a quick PR for KubeSpawner it's become clear that the better fix is probably on the JH API side than in KubeSpawner. Looks like the `/users` API doesn't return the `user_options` entry in the `spawners` table, which is already filled out with the user's `profile_list` selection. This would be the most flexible way to identify what sort of session the user is in instead of forcing a match against the container image, which is going to be a pain if you're not using a tag like `latest` or `stable`

Clearly I need to call it a day, sorry for all the noise.

**Comment by rkdarst** [comment]
I have a proposal... something I started a while ago but hasn't been
finished.

The JupyterHub API now exposes the "state" through the API (if you are
admin).  For KubeSpawner, this only has pod_name, but (for example)
ProfileSpawner includes the profile name in here.  This can be used
within the cull_idle script.  This is what I use to cull differently
depending on profile (actually, I have a hack from before the state
was always included... but it will be this once I get around to
updating).

KubeSpawner could be extended to include the profile name in state.
Then that can be accessed from the culler.  I have been meaning to
modify the culler script to clearly show where this should be added.

This could also be done somewhat local-only, by subclassing
KubeSpawner and overriding the get_state method.  But I think this
would be a useful upstream contribution to kubespawner (and really can
be added to most spawners...)  Or more generally, all the user options
could be saved and returned via the get_state method.  (While
technically this isn't state, it doesn't have to be used when
restoring state and can be _quite_ useful.

... though we could debate if it's better to add user_options to API
in addition to the state.  That would fix for every spawner at once,
but require storing those options (which isn't done now, and you could
argue is part of the state...)

@minrk, what do you think?

**Comment by rochaporto** [comment]
Reading your description it seems to provide exactly what i meant. Do you have a pointer to the code doing culling depending on the profile? I'm checking the wrap and profile spawners as well, probably something we'll need.

**Comment by vilhelmen** [comment]
Unfortunately, no. There's nothing really in place for this.

I think I'm still stuck waiting for some resources to build a demo JH cluster for some IEEE thing, so I'll start working on a patch to add configuration and we can iron out where and what the profile data is later (though I'm still for expanding the API data to include the missing column).

My plan is roughly a list of strings to index the spawner dict recursively to a dict with a max age and idle limit.

**Comment by vilhelmen** [comment]
I have a [rough (read: haven't tried running) solution](
https://github.com/vilhelmen/jupyterhub/blob/cull_config/examples/cull-idle/cull_idle_servers.py), but I have zero idea how to do configuration of the two new config variables (currently jammed in at the top) since command line arguments won't support the complexity we need. I could certainly use some guidance.

**Comment by rkdarst** [comment]
I just made #2613 with my hint for how to modify `cull_idle_servers.py` to implement custom logic, but it's the same as @vilhelmen has already done in his sample.  The trick is getting some useful state which *can* be used in the script, since by default kubespawner only includes the pod name there.  Which @vilhelmen is already working on in jupyterhub/kubespawner#327.

This hack could work in the meantime:

```
class KubeSpawnerWithProfileInfo(KubeSpawner):
    def get_state(self):
        state = super().get_state()
        state['culltime'] = self.culltime
        return state
```

Then include a `culltime` in the `kubespawner_override` in the `profile_list`.  It works its way to `cull_idle_servers.py` in the `server['state']` where my comment is.

Note: in theory this works, in practice I haven't tested at all - I use this with wrapspawner/batchspawner, which does include some info in `state` by default.

It would be nice if kubespawner would include the profile name in get_state by default.  @vilhelmen is already thinking about this in jupyterhub/kubespawner#327, but as said there it's not so easy.  Unfortunately kubespawner doesn't seem to save the profile name on the spawner object itself at all right now... giving us limited ability to add stuff.

**Comment by rkdarst** [comment]
I guess one fundamental question to @minrk is: should `state` be only used for the minimum needed to reinstantiate the spawner after the server has restarted, or can it include other useful information which can be used in, for example, cull_idle?

My thought is it's useful to say "a little bit can be included, but let's not get too excessive".

**Comment by yuvipanda** [comment]
> I guess one fundamental question to @minrk is: should `state` be only used for the minimum needed to reinstantiate the spawner after the server has restarted, or can it include other useful information which can be used in, for example, cull_idle?

Yeah, this is my understanding of what spawner state is to be used for. Maybe we should have a different endpoint if we wanna expose arbitrary state data? @minrk would definitely know better. Also cc @gsemet who added profile_list support to KubeSpawner

**Comment by vilhelmen** [comment]
This is why I was thinking returning the `user_options` db entry alongside state via the JH API would be good. It already records the user's `profile_list` selection, it should be simple enough to add extra management data.

**Comment by minrk** [comment]
Spawner state was *created* to reconstruct the Spawner across process restarts (e.g. container id for DockerSpawner, PID for local processes, etc.). That doesn't mean it's the only thing it can be used for.

I have mixed feelings about leaking internal implementation details to the API. I can certainly imagine in a deployment, seeing and making use of container IDs or PIDs, but it also shouldn't be leaked to e.g. *users*, only admins.

user_options is different from Spawner.state in this regard, since it is information used *as input* for the API (e.g. if you wanted to look at a running server and make another one with the same or similar inputs, you could pull user_options from one and make the appropriate request to start the other). State is different, since it's not an input to anything via a public API.

The clearest answer to the headline in this Issue is to use user_options for that. *However*, cases like ProfileSpawner add a layer of abstraction between the user_options and the actual value you're interested in, which makes things a bit more murky. Of course, the culler could know about profiles and make similar mappings from profile to cull settings, but that's probably going to involve duplicating that mapping somewhere. State is logical in that it's another arbitrary dict in which to dump things, but I'm not sure about making it available to the API.

**Comment by vilhelmen** [comment]
Thank you for your input, Min. I was somewhat hesitant about exposing more internal data, but I figured it would be ok-adjacent since it's read only. The heart of the issue seems to be that the culler is external, so why not tackle that head on!

We already have `Authenticator.refresh_user`, which runs "one in awhile" and lets us force logout a user (among other things). So, why not duplicate something like that over in `Spawner` that lets us gracefully* terminate a session. Leave it up to the user to define their logic, and provide a default for the current culler functionality via something like `Spawner.idle_timeout`/`Spawner.max_lifetime` that defaults to `None`/`None`.

*??? What's the current behavior for spawner shutdown? Do we save everything that was open? Are we relying on checkpoints? Either way, the culler already triggers a spawner shutdown, so it can't be any worse than the current behavior.

**Comment by rkdarst** [comment]
My small contribution:

> Spawner state was created to reconstruct the Spawner across process restarts ... That doesn't mean it's the only thing it can be used for.

+1.  Knowing relevant user config can be considered state that needs reloading.

>  State is logical in that it's another arbitrary dict in which to dump things, but I'm not sure about making it available to the API.

The current state is already in the API...

> State is different, since it's not an input to anything via a public API.

and the state in the API is admin-only

> However, cases like ProfileSpawner add a layer of abstraction between the user_options and the actual value you're interested in...

I think if you're using ProfileSpawner, then it's probably the admin's responsibility to deal with that layer.  At least for now.  We just pass it through with no extra work or complexity on our part.  ProfileSpawner already does a get_state on the child to pass whatever your child spawner wants through, so I think it's almost as good as can be.

So you can guess what I think: we have it, let's use it rather than re-designing something in parallel.  We don't need to care if a spawner puts extra stuff there, but it has to be able to use only the right info to reconstruct itself.  This would include reconstructing timeouts, etc. - which of course could change in the default profile.  It's up to the admin to decide if they put the profile or computed values in there.

**Comment by vilhelmen** [comment]
I took an hour or two and cobbled together a potential solution for internalizing the culler. I won't be upset if we decide not to go in this direction and throw it out.
